### PR TITLE
+str #19020 reduce combinator

### PR DESCRIPTION
**Pull request by agolubev**
Ref #19020

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 1]
```diff
@@ -974,3 +974,37 @@ private[stream] class DropWithin[T](timeout: FiniteDuration) extends SimpleLinea
...
+    override def toString = s"Reduce.Logic(aggregator=$aggregator)"
+    var aggregator: Option[T] = None
+
+    setHandler(in, new InHandler {
+      override def onPush(): Unit = {
+        aggregator = Some(aggregator match {
+          case Some(agg) ⇒ f(agg, grab(in))
+          case None      ⇒ grab(in)
+        })
+        pull(in)
```
Comment (hepin1989) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: shouldn't this be better performance via a `first` check or switch the `InHandler` after the first element arrived?

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Sounds like a great idea, hepin

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala [hunk 2]
```diff
@@ -710,10 +710,30 @@ trait FlowOps[+Out, +Mat] {
...
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   *
+   * See also [[FlowOps.fold]]
+   */
+  def reduce[T >: Out](f: (T, T) ⇒ T): Repr[T] =
+    via(new Reduce[T](f).withAttributes(name("reduce")))
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: remove `.withAttributes(name("reduce"))` and add to Reduce: `override def initialAttributes: Attributes = DefaultAttributes.reduce`

File: akka-docs-dev/rst/java/stream-cookbook.rst [hunk 3]
```diff
@@ -133,8 +133,8 @@ stream cannot continue without violating its resource bound, in this case
 By extracting the parts specific to *wordcount* into
 
 * a ``groupKey`` function that defines the groups
-* a ``foldZero`` that defines the zero element used by the fold on the substream given the group key
-* a ``fold`` function that does the actual reduction
+* a ``map`` map each element to value that is used by the reduce on the substream
+* a ``reduce`` function that does the actual reduction
```
Comment (drewhk) [inline akka-docs-dev/rst/java/stream-cookbook.rst]: Is this change correct? A reduce needs at least two elements, while a wordcount is obviously doable with only one word of a kind.

Comment (drewhk) [inline akka-docs-dev/rst/java/stream-cookbook.rst]: ah sorry I mixed up with `reduceLeft/Right`

**Comment by rkuhn** [comment]
closed by github due to deletion of the release-2.3-dev branch—please open a new PR against master

**Pull request closed without merging**